### ISSUE TITLE
Training on GPU and CPU with the same code differs significantly.

### ISSUE THREAD

**Issue by Mojusko** [tensorflow/tensorflow#6153]
I am training a TF network on two machines. It is the same code, which is basically a modification of the MNIST example. In one case, I use GPU on Amazon Cloud, in the other I use a standard CPU Intel Xeon. I have the same version of TF, the same code, the same parameters, the same packages the and same Ubuntu distribution.

In one case on GPU the learning process is significantly different, and particularly slower. It reaches only 0.60 performance while CPU one reaches 0.90. Increasing learning rate by 10 I managed to get the GPU result to 0.90 too. Have you encountered this problem before? Is this intended?

**Comment by yaroslavvb** [comment]
This may be WAI depending on how fast your GPU implementation runs. It's easy to end up with scenarios where GPU version runs 10x slower. The thing to look at is the performance after same number of iterations

**Comment by Mojusko** [comment]
@yaroslavvb, No I have a rather sophisticated validation stopping but in essence it boils down to iteration. So it is not the time I am concerned with it is really the computation. Could it be that there are different floating point standards leading to this, or the Adam optimizer behaves differently on GPU than on CPU. As I said raising the learning rate by 10 led to the same behavior.

**Comment by llhe** [comment]
@Mojusko Could it caused by random initialization for the variables?

**Comment by Mojusko** [comment]
No, the same numpy seed leeds to the same behavior, and if even this is not the relevant seed, it is too consistent to be caused by random fluctuations.

**Comment by michaelisard** [comment]
@Mojusko I'm sorry, this is frustrating for you. Unfortunately it's hard to guess from this level of detail what the root cause might be. I'm afraid the best next step might be for you to divide and conquer to see where the computations are differing, i.e. fetch tensors from the middle of your model, and locate the first place where the CPU/GPU results start to differ substantially.

One place you might focus on is reductions, since the order of reduction is known to be very different on CPU/GPU. Is it possible that your model has tensors laid out in some way that CPU reduction order consistently leads to faster training? (If so, that in itself might be interesting to understand from a model perspective.)

Sorry not to have more concrete solutions.

**Comment by Mojusko** [comment]
@michaelisard thank you for suggestions. It is not frustrating that much. I posted it here because could not find any similar answer on the net, and was wondering if someone else had experience with this (as I have specific machine in mind AWS p2 instance), and whether this is possibly a normal behavior. I will dive into further investigations.

**Comment by michaelisard** [comment]
Closing this issue for now but please reopen if you get a more specific bug report.

**Comment by fvisconti** [comment]
Hi all,

I have exactly the **opposite** behaviour.
My net is quite simple:
```python
model = Sequential()
    model.add(Conv2D(32, (3, 3), input_shape=input_shape))
    model.add(Activation('relu'))
    model.add(MaxPooling2D(pool_size=(2, 2)))
    model.add(Conv2D(32, (3, 3)))
    model.add(Activation('relu'))
    model.add(MaxPooling2D(pool_size=(2, 2)))
    model.add(Conv2D(64, (3, 3)))
    model.add(Activation('relu'))
    model.add(MaxPooling2D(pool_size=(2, 2)))
    model.add(Flatten())
    model.add(Dense(64))
    model.add(Activation('relu'))
    model.add(Dropout(0.5))
    model.add(Dense(1))
    model.add(Activation('sigmoid'))

    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
```
same inputs, running on two systems, CPU (Intel Core i5, compiled from source with SIMD support) and GPU (Nvidia Tesla K20).

GPU run is significantly faster (as expected, and I'm quite surprised to read this issue here); what is surprising is that after the first epoch on **CPU I have**
```4951s 566ms/step - loss: 0.7037 - acc: 0.4903  - val_loss: 0.6931 - val_acc: 0.5000```
while **on GPU**
```2892s 331ms/step - loss: 0.8896 - acc: 0.7267  - val_loss: 0.2933 - val_acc: 0.8777```.

How can I investigate in this?

**Comment by fvisconti** [comment]
Found this: https://stackoverflow.com/a/43264395/827818
I added these lines: 
```python
    np.random.seed(1)
    from tensorflow import set_random_seed
    set_random_seed(2)
```
In order to exclude randomness issues.

**Comment by annemenini** [comment]
@Mojusko : At the end, was the seed the problem? 
I am also seeing a similar problem (Ubuntu, python 3.5, TF1.8).
Same code, same data, same machine, just switching between GPU and CPU execution by adding os.environ["CUDA_VISIBLE_DEVICES"]="" gives fundamentally different results (the loss function differs by several orders of magnitude at the end of the first run of the first epoch).

I gave names to all the variables and operations in order to track what was the first operation in the graph where the results differ between CPU and GPU execution. I could actually identify it. In my case, it is a tf.multiply (pixel-wise multiplication of complex64 Tensors). I can see that the inputs of that operation are identical (+/- numerical precision), while the outputs are really different: the magnitude of the complex output is consistent but the phase is very different (the real and imaginary parts are both very different). I can tell that the CPU output is the right one.
Unfortunately, I am not able to reproduce the problem in a unit test. If I feed the same inputs to tf.multiply in a unit test, the CPU and GPU outputs are consistent. I am wondering if the CPU/GPU execution may be affecting the order in which the graph is build/executed, and if there may be interferences somehow.

Regarding the seeds, they are all set (graph wise and operation wise), but anyway, the diverging operation does not depend on a seed in this case.

If I switch the environment (same machine, same OS, same code, same data, but python 2.7, TF 1.4), the problem disappears, i.e. the CPU and GPU outputs are consistent.

Any idea or debugging experiment suggestion is very welcome.

**Comment by zhanghaoie** [comment]
@annemenini I have also seen this problem, have you got any solutions?
